### ISSUE TITLE
Upgrade WindowsAzure.Storage to Microsoft.Azure.Storage.*

### ISSUE THREAD

**Issue by kawazoe** [Azure/azure-functions-durable-extension#952]
I recently noticed, by working on a durable function that need newer features of the Azure blob storage client, that the durable function framework still use the older WindowsAzure.Storage nuget package.

Last year, the Azure Storage team split it into multiple packages and stopped maintaining the old package. Considering that v2 is still in preview, I think that now is a great opportunity to update to the newer Microsoft.Azure.Storage.* set of packages.

In the meantime, I'm stuck having to deal with both packages in my project since I need the newer features that got released in v11 which is a bit of a pain.

**Comment by cgillum** [comment]
We are looking to move to the newer storage libraries soon in order to get some badly needed reliability fixes, but you shouldn't be blocked from using the new libraries in your own application logic. Can you elaborate on the issues that you're dealing with in the current setup?

**Comment by kawazoe** [comment]
It’s not that I am locked out of using them, but I find annoying that I have to include both libraries in my code. It makes it messy to write storage related code as well since every class will get suggested twice by intellisense. Again, it’s just a small annoyance but I was wondering if you were aware of / working on it considering the updated packages were released over a year ago.

**Comment by cgillum** [comment]
Makes sense - thanks for the additional context. We were aware of the updated packages but couldn't easily use them because of the impact it would have on existing apps. We're now at a point where we feel we (the Functions host and related extensions) *need* to move, but it will require some coordination across multiple projects. I don't have an ETA, but we are in the planning phase for this.

**Comment by slarti-zak** [comment]
Any news here?

We are using the durable task extension but it is still referencing the deprecated WindowsAzure.Storage. We are even getting Azure Advisor Notifications about unsupported Version of the Storage Framework:

> Update Storage Client Library
> Minimum Recommended Version - 10.0.1
> Current Version - 9.3.1

But we are unable to fix this as it is still referenced by the durable tasks in 2022.

**Comment by maQus123** [comment]
+1

A log entry similar to the one below occurs once a minute in our Functions Log Stream and we assume it causes costs increase on Functions Storage. The durable orchestrator function mentioned is configured to run once a month. Does anyone else experience same behavior? 

```
2022-02-16T10:06:29Z   [Information]   Request [a51a8a3d-77da-4239-8f5b-7863a18c3b1e] PUT https://{FuncStorage}.blob.core.windows.net/azure-webjobs-hosts/locks/{FuncProject}/{DurableOrchestrator}.Listener?comp=lease
x-ms-lease-action:acquire
x-ms-lease-duration:15
x-ms-proposed-lease-id:0277bdd2-e50b-4cb9-8635-94766e46705d
x-ms-version:2020-08-04
Accept:application/xml
User-Agent:azsdk-net-Storage.Blobs/12.9.0,(.NET 6.0.0-rtm.21522.10; Linux 5.4.81-microsoft-standard #1 SMP Thu Dec 3 23:47:24 UTC 2020)
x-ms-date:Wed, 16 Feb 2022 10:06:29 GMT
Authorization:REDACTED
client assembly: Azure.Storage.Blobs
```

**Comment by ynijsen** [comment]
This is getting a bit ridiculous. It's not that it's deprecated for only a couple of weeks. Keep the packages up-to-date please 
![Screenshot 2022-03-11 180833](https://user-images.githubusercontent.com/6556813/158339810-2617a266-bf80-445d-8ef3-699a6d6b235c.png)

**Comment by kawazoe** [comment]
So, this issue [have been unresolved for over 3 years now](https://github.com/Azure/durabletask/blob/main/src/DurableTask.AzureStorage/DurableTask.AzureStorage.csproj#L46). I haven't had the chance to use Azure Functions for a while, and I'm about to jump into a new project that would benefit from them. While, at the time, I wasn't too bothered by this, I also expected it to get fixed rather quickly.

I am now running increasingly concerned of the security risks that comes with not only running but actively designing for and depending on unsupported software like this.

What happens if someone finds a major security issue in WindowsAzure.Storage à-la-log4j and everyone needs to update to the latest version within hours to mitigate the risks?

When I created this issue, this dependency was already a year old. Now, it has been plain and simple *deprecated* and *unmaintained* for over a year and a half. If an issue like this is found, I would be shocked to see it fixed retroactively in that version of the library. In fact, I don't even know what kind of version number they would give this fix, considering Durable Functions aren't even using the latest *bugfix* in their minor.

It's sad because I want to love this tool, but with dependency management like this, I cannot in my right mind recommend them for my next projects. At least... not until https://github.com/microsoft/durabletask-netherite becomes the new standard.

**Comment by wsugarman** [comment]
I have started this migration in the DTFx, and am now beginning to port the changes into the Durable Functions extension. I suspect this new version will first enter prerelease given the rather large number of changes.

Related: https://github.com/Azure/durabletask/issues/516

**Comment by lilyjma** [comment]
Hi all, we released [Durable Functions extension v3 preview](https://www.nuget.org/packages/Microsoft.Azure.WebJobs.Extensions.DurableTask/3.0.0-rc.1) which uses the latest Azure Storage SDKs. (It also has huge cost benefits for users of the Azure Storage backend.) 

Please note that upgrading to DF extension v3 may a breaking change for some .NET in-process users.

More details in this [blog post](https://techcommunity.microsoft.com/t5/apps-on-azure-blog/preview-of-durable-functions-extension-v3-0-0/ba-p/4000452)

**Comment by wsugarman** [comment]
v2.7.0+ should currently support the same identity-based authentication approaches as v3+; they both rely on the `Microsoft.Extensions.Azure` package to create `TokenCredential` objects (defined in `Azure.Identity`) based on the user's settings. However, v3+ also uses the latest Azure Storage SDKs which integrate much more easily with the `Azure.Identity` package.

**Comment by chandlerkent** [comment]
@wsugarman thank you for the response! Your work on adding identity support to Azure Functions has been very beneficial to my company.

**Comment by lilyjma** [comment]
Closing this issue as completed since the GA version of Durable Functions extension v3 has been out for a while.
